Update user-facing error and CLI messages to align with the style guidelines (#78)
heroku/sf-functions-python: user-facing-messages.md

# Python Functions user-facing messages

## CLI help text

Note: End users mostly won't use the sf-functions-python CLI directly, since they will
instead use the sf CLI which wraps it - however, I'm including these for completeness.

### sf-functions-python --help

```term
usage: sf-functions-python [-h] {check,serve,version} ...

Salesforce Functions Python Runtime

options:

>   -h, --help            show this help message and exit

Author: Unfortunately the `--help` messages are auto-generated by the Python stdlib's `argparse` module, so we cannot capitalise these unless I manually implement the module's built-in functionality, which I'd prefer not to do (especially given most users won't be interacting with the `sf-functions-python` CLI directly, and only the `sf` CLI).

>   -h, --help            show this help message and exit

Reviewer: Gotcha, that's fine!

subcommands:
  {check,serve,version}
    check               Checks that a function project is configured correctly
    serve               Serves a function project via HTTP
    version             Prints the version of the Python Functions Runtime
```

### sf-functions-python check --help

```term
usage: sf-functions-python check [-h] <project-path>

positional arguments:
  <project-path>  The directory that contains the function

options:

>   -h, --help      show this help message and exit

Author: Same as https://github.com/heroku/sf-functions-python/pull/78/files#r1093758684

```

### sf-functions-python serve --help

```term
usage: sf-functions-python serve [-h] [--host HOST] [-p PORT] [-w WORKERS]
                                 <project-path>

positional arguments:
  <project-path>        The directory that contains the function

options:
  -h, --help            show this help message and exit
  --host HOST           The host on which the web server should bind (default:
                        localhost)
  -p PORT, --port PORT  The port on which the web server should listen
                        (default: 8080)
  -w WORKERS, --workers WORKERS
                        The number of worker processes (default: 1)
```

## Checking/running a valid function

In the context of the buildpack self-check:

```term
Function passed validation
```

In the context of the CLI start command:

```term
Starting sf-functions-python v0.4.0 in single process mode.

> INFO:     Started server process [81699]

Reviewer: I wondered that, but edited it anyway, just in case.  So no worries!

INFO:     Waiting for application startup.
INFO:     Application startup complete.

> INFO:     Uvicorn running on http://localhost:8080 (Press CTRL+C to quit)

Author: Same as https://github.com/heroku/sf-functions-python/pull/78/files#r1093760852

ERROR:    Application startup failed. Exiting.
```

### tests/fixtures/project_toml_file_missing

In the context of the production deploy self-check:

```term
Function failed validation: A project.toml file was not found at: /Users/emorley/src/sf-functions-python/tests/fixtures/project_toml_file_missing/project.toml
```

In the context of the CLI start command:

```term
Starting sf-functions-python v0.4.0 in single process mode.
INFO:     Started server process [81734]
INFO:     Waiting for application startup.
ERROR:    RuntimeError: Unable to load function: A project.toml file was not found at: /Users/emorley/src/sf-functions-python/tests/fixtures/project_toml_file_missing/project.toml

ERROR:    Application startup failed. Exiting.
```

### tests/fixtures/project_toml_invalid_toml

In the context of the production deploy self-check:

```term
Function failed validation: project.toml is not valid TOML: Expected '=' after a key in a key/value pair (at line 1, column 4)
```

In the context of the CLI start command:

```term
Starting sf-functions-python v0.4.0 in single process mode.
INFO:     Started server process [81736]
INFO:     Waiting for application startup.

> ERROR:    RuntimeError: Unable to load function: project.toml is not valid TOML: Expected '=' after a key in a key/value pair (at line 1, column 4)

Author: For this and the next one, the changes to the start of the error message look great, however I'm less sure about the trailing full stop at the end, only because we don't generate the error message that's appended - so the library might include some that already have a trailing full stop, which would mean a double full stop.

> ERROR:    RuntimeError: Unable to load function: project.toml is not valid TOML: Expected '=' after a key in a key/value pair (at line 1, column 4)

Reviewer: makes sense!  so ignore it.  thx for explaining :)

ERROR:    Application startup failed. Exiting.
```

### tests/fixtures/project_toml_invalid_unicode

In the context of the production deploy self-check:

```term

> Function failed validation: Could not read project.toml: UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte

Reviewer: ```suggestion
Function failed validation: Couldn't read project.toml: UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte.
```

```

In the context of the CLI start command:

```term
Starting sf-functions-python v0.4.0 in single process mode.

> INFO:     Started server process [81738]

Reviewer: ```suggestion
INFO:     Started server process [81738].
```

INFO:     Waiting for application startup.

> ERROR:    RuntimeError: Unable to load function: Could not read project.toml: UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte

Reviewer: ```suggestion
ERROR:    RuntimeError: Unable to load function: Couldn't read project.toml: UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte.
```

ERROR:    Application startup failed. Exiting.
```

### tests/fixtures/project_toml_salesforce_table_missing

In the context of the production deploy self-check:

```term

> Function failed validation: project.toml is missing the required '[com.salesforce]' table.

Reviewer: ```suggestion
Function failed validation: The project.toml file is missing the required '[com.salesforce]' table.
```

```

In the context of the CLI start command:

```term
Starting sf-functions-python v0.4.0 in single process mode.

> INFO:     Started server process [81740]

Reviewer: ```suggestion
INFO:     Started server process [81740].
```

These are cases we cannot catch using the self-check, as they only occur when the function is running.

### tests/fixtures/raises_exception_at_runtime

Server log:

```term
Starting sf-functions-python v0.4.0 in single process mode.
INFO:     Started server process [81743]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://localhost:8080 (Press CTRL+C to quit)
Traceback (most recent call last):
  File "/Users/emorley/src/sf-functions-python/salesforce_functions/_internal/app.py", line 79, in _handle_function_invocation
    function_result = await function(event, context)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/emorley/src/sf-functions-python/tests/fixtures/raises_exception_at_runtime/main.py", line 5, in function
    return 1 / 0
           ~~^~~
ZeroDivisionError: division by zero
invocationId=00DJS0000000123ABC-fa35b79dc4047411826d3c52071066b9 level=error msg="Exception occurred whilst executing function: ZeroDivisionError: division by zero"
```

### tests/fixtures/return_value_not_serializable

Server log:

```term
Starting sf-functions-python v0.4.0 in single process mode.

> INFO:     Started server process [81754]

Reviewer: ```suggestion
INFO:     Started server process [81754].
```

INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://localhost:8080 (Press CTRL+C to quit)

> invocationId=00DJS0000000123ABC-0baf07e3d0f724a9b944165a86e51f89 level=error msg="Function return value cannot be serialized: TypeError: Type is not JSON serializable: set"

Reviewer: ```suggestion
invocationId=00DJS0000000123ABC-0baf07e3d0f724a9b944165a86e51f89 level=error msg="Function return value can't be serialized: TypeError: Type is not JSON serializable: set."
```

Author: Same as https://github.com/heroku/sf-functions-python/pull/78/files#r1093776099 for the trailing full stop.

```

### Invalid CloudEvent payload (should not be possible to trigger this in practice)

Server log:

```term
Starting sf-functions-python v0.4.0 in single process mode.
INFO:     Started server process [81767]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://localhost:8080 (Press CTRL+C to quit)

> level=error msg="Could not parse CloudEvent: Data payload is not valid JSON: invalid literal: line 1 column 1 (char 0)"

Reviewer: ```suggestion
level=error msg="Couldn't parse CloudEvent: Data payload isn't valid JSON: invalid literal: line 1 column 1 (char 0)."
```
